tests/omprog-output-capture-mt.sh

```diff
@@ -0,0 +1,94 @@
+# Similar to the 'omprog-output-capture.sh' test, with multiple worker
+# threads on high load. Checks that the lines concurrently emmitted to
+# stdout/stderr by the various program instances are not intermingled in
+# the output file (i.e., are captured atomically by omprog) when 1) the
+# lines are less than PIPE_BUF bytes long and 2) the program writes the
+# lines in line-buffered mode. In this test, the 'stdbuf' utility of GNU
+# Coreutils is used to force line buffering in a Python program (see
+# 'omprog-output-capture-mt-bin.py' for alternatives). Also checks the
+# throughput of writing the output (measured in lines written per second)
+# against a rough expected minimum (simple performance regression test).
+
+number_of_messages=30000   # number of logs to send
+expected_throughput=10000  # minimum written lines per second (2 lines x log)
+
+if [[ "$(uname)" == "Linux" ]]; then
+    line_length=4095   # 4KB minus 1 byte (for the newline char)
+else
+    line_length=511   # 512 minus 1 byte (for the newline char)
+fi
+
+export command_line="/usr/bin/stdbuf -oL -eL $srcdir/testsuites/omprog-output-capture-mt-bin.py $line_length"
+
+. $srcdir/diag.sh init
+. $srcdir/diag.sh check-command-available stdbuf
+generate_conf
+add_conf '
+module(load="../plugins/imtcp/.libs/imtcp")
+module(load="../plugins/omprog/.libs/omprog")
+
+input(type="imtcp" port="'$TCPFLOOD_PORT'")
+
+template(name="outfmt" type="string" string="%msg%\n")
+
+main_queue(
+    queue.timeoutShutdown="30000"  # long shutdown timeout for the main queue
+)
+
+:msg, contains, "msgnum:" {
+    action(
+        type="omprog"
+        binary=`echo $command_line`
+        template="outfmt"
+        name="omprog_action"
+        queue.type="LinkedList"  # use a dedicated queue
+        queue.workerThreads="10"  # ...with many workers
+        queue.timeoutShutdown="30000"  # ...and a long shutdown timeout
+        closeTimeout="10000"  # wait enough for program to terminate
+        output=`echo $RSYSLOG_OUT_LOG`
+        fileCreateMode="0644"
+    )
+}
+'
+startup
+SECONDS=0
+tcpflood -m$number_of_messages
+
+# Issue some HUP signals to cause the output file to be reopened during
+# writing (not a complete test of this feature, but at least we check it
+# doesn't break the output).
+. $srcdir/diag.sh issue-HUP
+sleep .1
+. $srcdir/diag.sh issue-HUP
+sleep .1
+. $srcdir/diag.sh issue-HUP
+
+shutdown_when_empty
+wait_shutdown
+elapsed_seconds=$SECONDS
+
+line_num=0
+while IFS= read -r line; do
+    let "line_num++"
+    if [[ ${#line} != $line_length ]]; then
+        echo "intermingled line in captured output: line: $line_num, length: ${#line} (expected: $line_length)"
+        echo "$line"
+        error_exit 1
+    fi
+done < $RSYSLOG_OUT_LOG
+
+if (( $line_num != $(($number_of_messages * 2)) )); then
+    echo "unexpected number of lines in captured output: $line_num (expected: $(($number_of_messages * 2)))"
+    error_exit 1
+fi
+
+throughput=$(($line_num / $elapsed_seconds))
```

Comment:
this is a very nice test, but be warned that some of  the CI machines are **really** slow. This must not lead to false positives (we are already plagued too much with them and I am working very hard to reduce them). Maybe this should be put to an optional test (which we than can carry out in a dedicated environment).

Reply:
True... I included this performance check because my first implementation of the functionality was inefficient and the check helped me refine it. However, I already had to decrease the threshold for the expected throughput (from more than 30.000 lines/sec in my machine to 10.000 lines/sec) just because I observed failures in my Travis CI. Now I realise that with this reduction it doesn't make sense to keep this check, since it would not detect the potential performance degradations it intended to cover.

Besides, it doesn't seem a good idea to include this performance check along with the other purposes of the test: for a more appropriate performance test, HUPs should not be included, the lines emitted by the program should be more realistic (less than 4K of size, which is a lot), and others. Moreover, the precision of the check is very rough (seconds).

So I have removed this performance check from the test. As you say, it would need to be a different test in any case, and this kind of tests require a dedicated environment.